Gem/Code/Source/Components/UI/UiSettingsComponent.cpp

```diff
@@ -140,91 +115,132 @@ namespace MultiplayerSample
     {
     }
 
+    template<typename ValueType>
+    uint32_t UiSettingsComponent::GetRotatedIndex(
+        const AZStd::vector<AZStd::pair<ValueType, AZStd::string>>& valuesToLabels,
+        const ValueType& value, ToggleDirection toggleDirection)
+    {
+        const size_t totalValues = valuesToLabels.size();
+
+        uint32_t curIndex = 0;
+
+        // Loop through and look for the correct value
+        for (size_t index = 0; index < totalValues; index++)
+        {
+            if (value == valuesToLabels[index].first)
+            {
+                curIndex = aznumeric_cast<uint32_t>(index);
+                break;
+            }
+        }
+
+        switch (toggleDirection)
+        {
+        case ToggleDirection::Left:
+            return aznumeric_cast<uint32_t>((curIndex + (totalValues - 1)) % totalValues);
+        case ToggleDirection::Right:
+            return aznumeric_cast<uint32_t>((curIndex + 1) % totalValues);
+        default:
+            return curIndex;
+        }
+    }
+
     void UiSettingsComponent::OnGraphicsApiToggle(ToggleDirection toggleDirection)
     {
         // This list is expected to match the values in AZ::RHI::ApiIndex.
-        const char* labels[] =
+        const AZStd::vector<AZStd::pair<AZStd::string, AZStd::string>> valuesToLabels =
```

Comment:
Out of curiosity, no action needed here, but do we have a GHI for Atom to maintain this mapping in the future so that adding a new rendering backend doesn't require modifications to customer product code?

Reply:
Not exactly. I have a change that I can put up for the development branch to add an API to Atom to expose the set of valid rendering backends. I just didn't want to submit it until I move from working on stabilization to working on development. This code would then be able to use that API to get the proper subset of factory names.

That said, the names we would get would be a subset of "null", "dx12", "vulkan", "metal". I'd personally be OK with using those as the display values, but if we wanted fancier names like "DirectX 12", or to filter out "null", we'd still need to do some pretty equivalent label mapping and hard-coding here.

We could theoretically add another API to the factories to get display names separately from factory names, but even then we'd _still_ potentially want to filter out "null", and potentially use different display names (especially if we ever localized), so not sure this can be fully solved the way you'd like.

I'm pretty sure new rendering backends will get added pretty infrequently though.